chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -595,91 +532,165 @@ impl Tree {
+            update_secret: SecretVec::new(vec![0; cs.hash_len()]),
         }
     }
 
-    /// find the intersection between sender and self.my_pos
-    /// update the secrets along the path to root
+    /// find the intersection between sender and self.my_pos,
+    /// and decrypt the path secret.
     ///
     /// returns None if tree only have one leaf
     fn decrypt_path_secrets(
         &self,
         sender: LeafSize,
+        tree: &TreePublicKey,
         ctx: &[u8],
         path_nodes: &[DirectPathNode],
         leaf_private_key: &HPKEPrivateKey,
     ) -> Option<(ParentSize, SecretVec<u8>)> {
-        let leaf_len = self.leaf_len();
-        let ancestor = ParentSize::common_ancestor(sender, self.my_pos);
+        let leaf_len = tree.leaf_len();
+        let ancestor = ParentSize::common_ancestor(sender, tree.my_pos);
         match ancestor {
             None => {
                 // same leaf
                 let node_index = NodeSize::from(sender);
                 Some((
                     node_index.parent(leaf_len)?,
-                    self.cs
+                    tree.cs
                         .expand_label(
                             &leaf_private_key.marshal(),
                             vec![],
                             "path",
                             &[],
-                            self.cs.secret_size(),
+                            tree.cs.secret_size(),
                         )
                         .expect("invalid length"),
                 ))
             }
             Some(ancestor) => {
-                let my_node = if self.my_pos > sender {
-                    ancestor.right(leaf_len)
-                } else {
-                    ancestor.left()
-                };
                 let path_node = NodeSize::from(sender)
                     .direct_path(leaf_len)
                     .into_iter()
                     .zip(path_nodes.iter())
                     .find(|(n, _)| *n == ancestor)
                     .expect("invalid path nodes")
                     .1;
-                for (i, node_index) in resolve(&self.nodes, my_node).iter().copied().enumerate() {
-                    if let Some(private_key) = self.node_private_key(node_index, leaf_private_key) {
-                        let secret = self
-                            .cs
-                            .decrypt(private_key, ctx, &path_node.encrypted_path_secret[i])
-                            .expect("decrypt failed");
-                        return Some((ancestor, SecretVec::new(secret)));
-                    }
-                }
-                panic!("don't find private key to decrypt secret");
+                let pos = NodeSize::from(tree.my_pos)
+                    .direct_path(leaf_len)
+                    .iter()
+                    .position(|&p| p == ancestor)
+                    .expect("ancestor in direct path");
+                let secret = match pos.checked_sub(1) {
```

Comment:
maybe some comment why `pos.checked_sub(1)` and what those two cases represent (no ancestors for None?)?

Reply:
yes, added some comments now.